### PR TITLE
LOOP-3856: Cancel temp basal if user lowers max basal below current temp basal

### PR DESCRIPTION
**Pull request by rickpasetto**
This implements cancellation of temp basal if the max basal rate is lowered below the current temp basal

https://tidepool.atlassian.net/browse/LOOP-3856

LWPR: https://github.com/tidepool-org/LoopWorkspace/pull/658

### CODE REVIEW

**Comment by rickpasetto** [comment]
@ps2 Please check my work...if you think this is the right approach I'll try to add some unit tests.

File: LoopKit/DeviceManager/PumpManager.swift [hunk 1]
```diff
@@ -147,11 +147,13 @@ public protocol PumpManager: DeviceManager {
     ///   - result: A DoseEntry containing the actual delivery amount of the canceled bolus, nil if canceled bolus information is not available, or an error describing why the command failed.
     func cancelBolus(completion: @escaping (_ result: PumpManagerResult<DoseEntry?>) -> Void)
 
+    typealias EnactTempBasal = (_ unitsPerHour: Double, _ duration: TimeInterval, _ completion: @escaping (_ error: PumpManagerError?) -> Void) -> Void
```
Comment (ps2) [inline LoopKit/DeviceManager/PumpManager.swift]: See my comments on https://github.com/tidepool-org/Loop/pull/460 about these typealiases.

File: LoopKit/DeviceManager/PumpManager.swift [hunk 2]
```diff
@@ -147,11 +147,19 @@ public protocol PumpManager: DeviceManager {
     ///   - result: A DoseEntry containing the actual delivery amount of the canceled bolus, nil if canceled bolus information is not available, or an error describing why the command failed.
     func cancelBolus(completion: @escaping (_ result: PumpManagerResult<DoseEntry?>) -> Void)
 
+    /// This method type describes a way to check if a proposed basal value and duration is "valid" for the pump manager.
+    ///
+    /// - Parameters:
+    ///   - unitsPerHour: The temporary basal rate proposed to validate, in international units per hour
+    ///   - completion: A closure called after the command is complete
+    ///   - error: An optional error describing why the command failed
+    typealias ValidateTempBasal = (_ unitsPerHour: Double, _ completion: @escaping (_ error: Error?) -> Void) -> Void
```
Comment (ps2) [inline LoopKit/DeviceManager/PumpManager.swift]: Hmm, what does "valid" mean in this context?

File: LoopKit/DeviceManager/PumpManager.swift [hunk 3]
```diff
@@ -147,11 +147,19 @@ public protocol PumpManager: DeviceManager {
...
+
     /// Send a temporary basal rate command and handle the result
     ///
     /// - Parameters:
     ///   - unitsPerHour: The temporary basal rate to set
-    ///   - duration: The duration of the temporary basal rate.
+    ///   - duration: The duration of the temporary basal rate.  If you pass in a duration of 0, that cancels the Temp Basal
```
Comment (ps2) [inline LoopKit/DeviceManager/PumpManager.swift]: 👍 👍 👍

File: LoopKit/DeviceManager/PumpManager.swift [hunk 4]
```diff
@@ -170,7 +178,7 @@ public protocol PumpManager: DeviceManager {
     ///   - error: An error describing why the command failed
     func resumeDelivery(completion: @escaping (_ error: Error?) -> Void)
     
-    typealias SyncSchedule = (_ items: [RepeatingScheduleValue<Double>], _ completion: @escaping (Result<BasalRateSchedule, Error>) -> Void) -> Void
+    typealias SyncBasalRateSchedule = (_ items: [RepeatingScheduleValue<Double>], _ completion: @escaping (Result<BasalRateSchedule, Error>) -> Void) -> Void
```
Comment (ps2) [inline LoopKit/DeviceManager/PumpManager.swift]: 👍

File: LoopKit/DeviceManager/PumpManager.swift (hunk 2 above)
Comment (ps2) [reply LoopKit/DeviceManager/PumpManager.swift]: Seems like this is for TherapySettingsViewModel?  It should probably be moved over there. And I think the documentation should describe that this is a "precheck" for saving max temp basal. I.e. the host app is going to supply a closure for this, and based on the response, will proceed with saving max temp, or display an error.

File: LoopKit/DeviceManager/PumpManager.swift (hunk 3 above)
Comment (ps2) [reply LoopKit/DeviceManager/PumpManager.swift]: "cancels the Temp Basal" -> "cancels any currently running Temp Basal"

File: LoopKitUI/ViewModels/TherapySettingsViewModel.swift [hunk 5]
```diff
@@ -20,21 +20,24 @@ public class TherapySettingsViewModel: ObservableObject {
...
+    let syncBasalRateSchedule: (() -> PumpManager.SyncBasalRateSchedule?)?
+    let validateTempBasal: (() -> PumpManager.ValidateTempBasal?)?
     let sensitivityOverridesEnabled: Bool
     public var prescription: Prescription?
 
     public init(therapySettings: TherapySettings,
                 supportedInsulinModelSettings: SupportedInsulinModelSettings = SupportedInsulinModelSettings(fiaspModelEnabled: true, walshModelEnabled: true),
                 pumpSupportedIncrements: (() -> PumpSupportedIncrements?)? = nil,
-                syncPumpSchedule: (() -> PumpManager.SyncSchedule?)? = nil,
+                syncBasalRateSchedule: (() -> PumpManager.SyncBasalRateSchedule?)? = nil,
```
Comment (ps2) [inline LoopKitUI/ViewModels/TherapySettingsViewModel.swift]: The types of these args appear to be closures that return closures? Why the nesting?

**Review by ps2 — COMMENTED** [review]
I think there may need to be more changes; asked a few questions. Maybe they lead to changes or just answers.

File: LoopKitUI/ViewModels/TherapySettingsViewModel.swift (hunk 5 above)
Comment (rickpasetto) [reply LoopKitUI/ViewModels/TherapySettingsViewModel.swift]: I thought of this.  The ones that themselves return closures are strange, it is true, but remember that's the way we ensure clients of `TherapySettingsViewModel` get updated values.  In practice it is probably unlikely they will need an "updated version of the `SyncBasalRateSchedule` function, but following the pattern is probably a better thing to do.

File: LoopKit/DeviceManager/PumpManager.swift (hunk 2 above)
Comment (rickpasetto) [reply LoopKit/DeviceManager/PumpManager.swift]: If we move this there, we should probably also move the `SyncBasalRateSchedule` `typealias`. I'll do that too.

File: LoopKitUI/ViewModels/TherapySettingsViewModel.swift (hunk 5 above)
Comment (ps2) [reply LoopKitUI/ViewModels/TherapySettingsViewModel.swift]: If the extra complexity is not needed, it should probably be simplified. Seeing the double abstraction implies to the reader that it is necessary.

File: LoopKitUI/ViewModels/TherapySettingsViewModel.swift [hunk 6]
```diff
@@ -13,28 +13,42 @@ import SwiftUI
...
     public typealias SaveCompletion = (TherapySetting, TherapySettings) -> Void
+   
+    /// This method type describes a way to "precheck" before saving max temp basal.  The host app is going to supply a closure for this, and based on the
+    /// response, either proceed with saving max temp, or display an error.
+    ///
+    /// - Parameters:
+    ///   - unitsPerHour: The temporary basal rate proposed to validate, in international units per hour
+    ///   - completion: A closure called after the command is complete
+    ///   - error: An optional error describing why the command failed
+    public typealias ValidateMaxTempBasal = (_ unitsPerHour: Double, _ completion: @escaping (_ error: Error?) -> Void) -> Void
```
Comment (ps2) [inline LoopKitUI/ViewModels/TherapySettingsViewModel.swift]: This is a hard one to name... keep trying to find something that covers the intent better. Maybe maxTempBasalSavePreflight?  The current name reads like we're trying make sure the passed max temp is valid. But yeah, it's tough.

**Review by ps2 — APPROVED** [review]
LGTM. Just a couple suggestions; one for naming, and the other that I do think it would be good to simplify call signatures if there is unneeded flexibilty.

File: LoopKitUI/ViewModels/TherapySettingsViewModel.swift (hunk 5 above)
Comment (rickpasetto) [reply LoopKitUI/ViewModels/TherapySettingsViewModel.swift]: Making such a change is unrelated to this PR, though, I think.

File: LoopKitUI/ViewModels/TherapySettingsViewModel.swift (hunk 6 above)
Comment (rickpasetto) [reply LoopKitUI/ViewModels/TherapySettingsViewModel.swift]: It's ugly to be sure... and probably a code smell.  I'm not a huge fan of the term "Preflight" (it's unusual) but I would like to get this PR in.

**Pull request merged**